### ISSUE TITLE
Warnings generated for mut parameters

### ISSUE THREAD

**Issue by bbarker** [alecdotninja/tailcall#7]
```rust
fn add_iter<'a, I>(mut int_iter: I, accum: i32) -> i32
where
    I: Iterator<Item = &'a i32>,
{
    match int_iter.next() {
        Some(i) => add_iter(int_iter, accum + i),
        None => accum,
    }
}
```

With 1.51.0 nightly (783bc43c6 2021-01-20), this code emits a warning:

```none
warning: variable does not need to be mutable
  --> tests/correctness.rs:18:20
   |
18 | fn add_iter<'a, I>(mut int_iter: I, accum: i32) -> i32
   |                    ----^^^^^^^^
   |                    |
   |                    help: remove this `mut`
   |
   = note: `#[warn(unused_mut)]` on by default
```

If I remove `mut` for `int_iter`, I get:

```none
error[E0596]: cannot borrow `int_iter` as mutable, as it is not declared as mutable
  --> tests/correctness.rs:22:11
   |
18 | fn add_iter<'a, I>(int_iter: I, accum: i32) -> i32
   |                    -------- help: consider changing this to be mutable: `mut int_iter`
...
22 |     match int_iter.next() {
   |           ^^^^^^^^ cannot borrow as mutable
```

Looking at the macro-generated code:

```rust
fn add_iter_exp<'a, I>(mut int_iter: I, accum: i32) -> i32
where
    I: Iterator<Item = &'a i32>,
{
    tailcall::trampoline::run(
        #[inline(always)]
        |(mut int_iter, accum)| {
            (tailcall::trampoline::Finish({
                match int_iter.next() {
                    Some(i) => return (tailcall::trampoline::Recurse((int_iter, accum + i))),
                    None => accum,
                }
            }))
        },
        (int_iter, accum),
    )
}
```

The warning disappears if the generated code has the first `mut` removed:

```rust
fn add_iter_exp<'a, I>(int_iter: I, accum: i32) -> i32
```

**Comment by bbarker** [comment]
It was suggested that we may want to look at [func_wrap](https://docs.rs/func_wrap) (or its source code) to get a better handle on the situation, since:

> Basically, mut is only one such example; imagine somebody using ref, or even _

> So you need to map a copy of the input sequence of patterns whereby each non-simple pattern is replaced by one:
`ref x: …, _: …, (y, z): …, foo: …,`

> becomes
`arg0: …, arg1: …, arg2: …, foo: …,`

**Comment by alecdotninja** [comment]
This was fixed via #11 which was released in [v0.1.5](https://crates.io/crates/tailcall/0.1.5). Thank you for all of your help here, @bbarker !
